design/one-pager-resource-usages.md

```diff
@@ -0,0 +1,346 @@
+type: Opaque
+data:
+  config: |
+    provider: AWS
+    region: aws_region
+    aws_access_key_id: aws_access_key_id
+    aws_secret_access_key: aws_secret_access_key
+```
+
+Inflexibility is not the only limitation of the contemporary connection secret
+pattern. Many external resources either generate or expect to be provided with a
+set of superuser credentials at creation time. Crossplane passes these superuser
+credentials on to any claimant, going against typical security best practices.
+`KubernetesCluster` claims, for example, will be provided with the cluster
+administrator's credentials whether they need them or not. Azure storage
+containers are another example of this mispattern. Crossplane distributes the
+storage account key to all claimants, contrary to the recommendation of the
+[Azure storage documentation]:
+
+> Your storage account key is similar to the root password for your storage
+> account. Always be careful to protect your account key. Avoid distributing
+> it to other users, hard-coding it, or saving it anywhere in plaintext that
+> is accessible to others.
+
+Azure recommends instead granting access via [Azure service principal], a
+distinct resource with its own credentials that may be granted access to storage
+containers and many other Azure resources. [GCP service accounts] and [AWS IAM
+users] are analogous to service principals in their respective clouds. Creating
+an AWS bucket transparently creates an IAM role that is granted access to the
+bucket. This existence of this IAM role is not modelled in the Kubernetes API.
+GCP buckets do not create a service account, but require one to be created
+manually outside of Crossplane in order to grant bucket access. Storage buckets
+happen to be the only class of resource currently supported by Crossplane that
+use such 'robot' accounts to grant access, but this is the prevalent pattern
+amongst most resources provided by the big three clouds. Hypothetical
+`DynamoDB`, `CloudDatastore`, `CloudSpannerInstance`, etc managed resources
+would also require such accounts.
+
+## Goals
+
+The design proposed by this document intends to:
+
+* Allow resource claimants to specify the format in which they require
+  connection details to be presented.
+* Avoid distributing superuser credentials where consumer scoped credentials
+  would be more appropriate.
+* Establish foundations for modelling the consumption of a managed resource in
+  order to eventually ensure network level connectivity between.
+
+Note that while this document may lay the foundation for network level
+connectivity constructs, it defers discussing that problem space in detail.
+
+## Proposed Design
+
+This document proposes:
+
+* That the existing resource class and managed resource kinds become cluster
+  scoped, rather than namespaced.
+* The introduction of cluster scoped, 'resource secrets' aligned with existing
+  managed resource kinds, for example `RDSInstanceSecret`, or
```

Comment:
It worth to call out here that affinitizing credentials information with a `resource` maybe not (always) correct representation and may require additional crossplane models. For example, `CloudSQLInstance` - the password value is *NOT* property of the `instance` but rather a property of the `instance user`, which there could be many of, i.e. a single `CloudSQLInstance` may have multiple `users`. I believe this is `RDSInstance` add `MySQLServerInstance` have a very similar paradigm.

Reply:
That's a good point

In some cases such as CloudSQL instances it appears that a database user, which exists as a distinct object in the GCP API, is created as an implicit side effect of creating the CloudSQL instance. In others such as the Azure and AWS SQL instances an user is created implicitly but exists only at the SQL level, not modeled in the cloud provider API. In other resources, such as all the Redis managed resources that support authentication, there is no concept of a "user" but only a shared secret that is associated directly with the instance.

So in most cases it may make sense to model the "master" credentials of a resource as a distinct `RDSUser`, or `CloudSQLUser`, (etc) resource in Crossplane, then somehow indicate that that user is the "master" user that should be used for creating databases, other users, etc.